### PR TITLE
Enable and fuse the Inkling MoE gate epilogue on Intel XPU

### PR DESCRIPTION
**Pull request by jmunetong**
## Enable and fuse the Inkling MoE gate epilogue on Intel XPU

**The Inkling MoE gate epilogue — everything between the 258-column gate GEMM and the MoE
dispatch — does not run on XPU at all today, and once it does, the path it lands on is
22–64× slower than necessary. Three independent commits: unblock, correct, optimize. CUDA and
ROCm are unchanged.**

### 1. Enablement: the dispatch predicate admits XPU

`sigmoid_gate_topk_renorm` gated its CUDA JIT path on `torch.version.hip is None`, which is also
true on XPU, so **all 64 MoE layers took the CUDA path** and died on `assert logits.is_cuda`
(`inkling_gate_topk_renorm.py:77`); fixed by gating on **`.is_cuda and torch.version.hip is
None`** — a conjunction, not a swap, because either predicate alone still admits XPU or ROCm.

### 2. Correctness: the Triton renorm returns NaN on underflow

`_sigmoid_gate_topk_renorm_kernel` formed weights as `sigmoid(x) / Σ sigmoid(x)`, and fp32
sigmoid flushes to zero below x ≈ −104, so once all `K + S` active logits are that negative the
denominator is exactly 0 and **every routed weight comes back NaN**; replaced with
`exp(lp − logsumexp(lp))`, `lp = logsigmoid(x)`, which is algebraically identical and already
what the eager reference uses (`inkling_common/moe.py:140`).

### 3. Performance: give the unified router an Inkling epilogue

The fallback ranks the `[T, 258]` fp32 gate row with `tl.topk`, a bitonic sort over all 256
routed columns that is pathological on Xe, whereas `_router_triton_kernel` already ranks the
cheap way with masked-max passes — so this parameterizes that kernel's **epilogue** with four
constexprs (`EPILOGUE`, `SHARED_SINK`, `HAS_GLOBAL_SCALE`, `RETURN_PACKED`) rather than adding
a kernel.

Device time (profiler self-time), same process, interleaved, min of 5 reps, Arc Pro B60:

| µs | T=1 | T=8 | T=64 | T=512 | T=4096 | T=8192 |
|---|---|---|---|---|---|---|
| `tl.topk` | 58.81 | 58.93 | 62.20 | 181.07 | 641.57 | 1237.55 |
| router | 2.43 | 2.46 | 2.77 | 3.40 | 15.64 | 28.02 |
| **speedup** | **24.2×** | 24.0× | 22.5× | **53.3×** | 41.0× | 44.2× |

### Correctness

Verified against an fp64 oracle (`rtol=atol=2e-3`): 12 shapes × 6 token counts **pass on all
12, max abs 3.6e-7, zero index flips**, including ties, a dominating shared sink, and a
non-multiple-of-`BLOCK` tail at T=37. **44 existing `moe_fused_gate` configs stay bit-identical
in weights and ids** with device time unchanged, since `LOGSIGMOID_SINK` lives entirely behind
the `EPILOGUE` / `SHARED_SINK` constexprs (both 0 for every existing caller) — so the result
carries to CUDA structurally.

### Tests

Adds `test/registered/xpu/test_inkling_gate_epilogue.py` — 9 cases / 33 subtests, ~14 s on an
Arc Pro B60, covering both dispatch branches against the fp64 oracle plus the invariant, packed
mode, ties, a dominating sink, a `BLOCK` tail, and `shared_sink == 0` non-regression; XPU CI
only per the `register_xpu_ci` BKM. Each case fails without its fix — reverting all three gives
21 failed / 4 passed — and the fp64 oracle is the gate because **the sum invariant alone does
not catch a wrong carry.**

### Test plan

```sh
WT=<this branch>/python
# CI test added here
PYTHONPATH=$WT ZE_AFFINITY_MASK=4,5,6,7 python -m pytest test/registered/xpu/test_inkling_gate_epilogue.py
# fp64 oracle + device time, 12 shapes x 6 token counts
PYTHONPATH=$WT ZE_AFFINITY_MASK=4,5,6,7 python bench_gate_epilogue.py --out after.json
# the A/B that produces the speedup table (bench alone cannot: the JIT env var is inert on XPU)
PYTHONPATH=$WT ZE_AFFINITY_MASK=4,5,6,7 python verify/ablate_stages.py --repeats 5 --iters 400
# 44-config bit-exactness at shared_sink == 0
PYTHONPATH=$WT ZE_AFFINITY_MASK=4,5,6,7 python verify/nonreg_moe_fused_gate.py --check golden_moe_fused_gate.pt
```

Intel Arc Pro B60, torch 2.12.0+xpu, triton 3.7.1. `ablate_stages.py` reports the kernel that
actually ran (`top=`), so the dispatch branch is proven rather than inferred.

---
### CI States

Latest PR Test (Base): :x: [Run #33813301137](https://github.com/sgl-project/sglang/actions/runs/33813301137)
Latest PR Test (Extra): :x: [Run #33813300723](https://github.com/sgl-project/sglang/actions/runs/33813300723)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33813301032](https://github.com/sgl-project/sglang/actions/runs/33813301032)

### CODE REVIEW

**Comment by jmunetong** [comment]
**Verdict: measured on Intel XPU at tp=4, this PR is worth +8.29% output tok/s and −7.85% TPOT with no resolved change in TTFT (p = 0.37) — and it is essentially the whole of the measured end-to-end effect.**

### End-to-end — n = 12 per arm (6 interleaved sessions × 2 unprofiled reps; p computed on session means, n = 6)

| metric | gate off (`_streaming_topk_kernel`) | gate on (`_router_triton_kernel`) | Δ | % | p |
|---|---|---|---|---|---|
| Output token throughput (tok/s) | 35.20 ± 0.84 | **38.12 ± 0.93** | **+2.92** | **+8.29%** | **0.0002** |
| TPOT (ms, mean) | 27.76 ± 0.68 | **25.58 ± 0.63** | **−2.18** | **−7.85%** | **0.0001** |
| Median ITL (ms) | 27.64 ± 0.72 | **25.41 ± 0.57** | **−2.23** | **−8.06%** | **0.0002** |
| TTFT (ms, mean) | 682.07 ± 18.81 | 687.41 ± 28.17 | +5.34 | +0.78% | 0.37 — **not resolved** |

### Per-op device time, decode per step — n = 6 sessions/arm; 🔴 = buckets this PR computes differently, decided by kernel identity and the `calls` column (router kernel replaced 1:1; `elementwise`/`memcpy`/`other` launches fused away)

| bucket | gate off | gate on | Δ | % | noise | calls |
|---|---|---|---|---|---|---|
| 🔴 **`moe_router_topk`** | **3.663** | **0.041** | **−3.622** | **−98.9%** | ±0.289 | 8 → 8 |
| 🔴 **`elementwise`** | 0.205 | 0.082 | −0.122 | −59.5% | ±0.014 | **93.8 → 25.8** |
| 🔴 **`memcpy`** | 0.101 | 0.025 | −0.076 | −75.2% | ±0.005 | **18.8 → 6.8** |
| 🔴 **`other`** | 0.043 | 0.035 | −0.008 | −18.6% | ±0.003 | **20.1 → 16.1** |
| **NON-COLLECTIVE TOTAL** | **8.241** | **4.512** | **−3.729** | **−45.2%** | ±0.458 | **301.9 → 217.9** |
| `comm_allreduce` | 21.240 | 23.349 | +2.108 | +9.9% | ±7.302 | 13 → 13 — **not resolved**, not a regression |

Prefill: `moe_router_topk` goes 3.497 → 0.075 ms, **−3.422 (−97.9%)** against a ±0.017 bar — but that is only **−0.53% of the prefill device total**, which is why TTFT shows nothing.

Attribution: the A/B is both-optimizations-vs-neither from **one binary, env-gated, no rebuild between arms**. The companion q-norm PR #37323 is **−0.011%** of prefill device time and **+0.0012 ms/step (p = 0.26, not resolved)** in decode, so the end-to-end effect above is attributable to this PR.

Environment: Intel Arc Pro B60, tp=4, oneAPI 2026.1, 6-layer reduced Inkling checkpoint, `--attention-backend triton`, `--page-size 64`, decode CUDA graphs on.
Exact repro (env vars, aggregation, gotchas) lives in the Reproduce section of our internal report `AB_gate_plus_norm_tp4.md` plus its `sweep_run2.sh`; happy to paste either here on request.

**Comment by jmunetong** [comment]
@airMeng could you look into this PR when you get a chance?

**Review by airMeng — COMMENTED** [review]
I think these are huge changes to CUDA/ROCm as well. Can we contribute these optimized triton functions to sgl-kernel-xpu then we can dispatch to the optimized path only for XPU here?
